File: src/main/java/xyz/novaserver/regionmusic/RegionPlayer.java
package xyz.novaserver.regionmusic;

import org.bukkit.SoundCategory;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import xyz.novaserver.regionmusic.task.PlayerMusicTask;
import xyz.novaserver.regionmusic.task.PlayerStopTask;

public class RegionPlayer {
    private final Plugin plugin;
    private final Player player;

    private SoundRegion region;
    private PlayerMusicTask musicTask;
    private PlayerStopTask stopTask;

    public RegionPlayer(Player player, Plugin plugin) {
        this.player = player;
        this.plugin = plugin;
    }

    public void playSound() {
        // Play custom music
        player.playSound(player.getLocation(), region.getSound(), region.getCategory(), region.getVolume(), region.getPitch());
    }

    public void stopSound() {
        player.stopSound(region.getSound(), region.getCategory());
    }

    public void stopGameMusic() {
        plugin.getConfig().getStringList("stop-sounds").forEach(sound -> player.stopSound(sound, SoundCategory.MUSIC));
    }

    public Player getPlayer() {
        return player;
    }

    public SoundRegion getRegion() {
        return region;
    }

    public void setRegion(SoundRegion region) {
        this.region = region;
    }

    public PlayerMusicTask getMusicTask() {
        return musicTask;
    }

    public void createMusicTask(Plugin plugin) {
        musicTask = new PlayerMusicTask(plugin, this);
        musicTask.schedule();
    }

    public PlayerStopTask getStopTask() {
        return stopTask;
    }

    public void createStopTask(Plugin plugin) {
        stopTask = new PlayerStopTask(plugin, this);
        stopTask.schedule();
    }
}

File: src/main/kotlin/xyz/novaserver/regionmusic/data/RegionPlayer.kt
package xyz.novaserver.regionmusic.data

import org.bukkit.Bukkit
import org.bukkit.SoundCategory
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitTask

class RegionPlayer(
    private val plugin: Plugin,
    private val player: Player
) {
    private var musicTask: BukkitTask? = null
    private var stopTask: BukkitTask? = null
    var event: SoundEvent = SoundEvent.NO_EVENT

    fun playSound() {
        player.playSound(player.location, event.sound, event.category, event.volume, event.pitch)
    }

    fun stopSound() {
        player.stopSound(event.sound, event.category)
    }

    fun stopGameMusic() {
        plugin.config.getStringList("stop-sounds").forEach { player.stopSound(it, SoundCategory.MUSIC) }
    }

    fun createMusicTask() {
        musicTask = Bukkit.getScheduler().runTaskTimerAsynchronously(plugin,
            Runnable { this.playSound() },1, event.length)
    }

    fun createStopTask() {
        stopTask = plugin.server.scheduler.runTaskTimerAsynchronously(plugin, Runnable { this.stopGameMusic() },
                20, plugin.config.getLong("stop-period", 80))
    }

    fun cancelTasks() {
        if (musicTask != null && !musicTask!!.isCancelled) {
            musicTask?.cancel()
        }
        if (stopTask != null && !stopTask!!.isCancelled) {
            stopTask?.cancel()
        }
    }
}

